ucfg_test.go

```diff
@@ -61,3 +68,284 @@ func TestFlattenKeys(t *testing.T) {
 		})
 	}
 }
+
+func TestDetectDirectCyclicReference(t *testing.T) {
+	tests := []struct {
+		title    string
+		cfg      map[string]interface{}
+		expected string
+	}{
+		{
+			title: "direct cyclic reference",
+			cfg: map[string]interface{}{
+				"top.reference": "${top.reference}",
+			},
+		},
+		{
+			title: "compound reference",
+			cfg: map[string]interface{}{
+				"top.reference": "hello ${top.reference}",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.title, func(t *testing.T) {
+			c, err := NewFrom(test.cfg, opts...)
+
+			config := &struct {
+				TopReference string `config:"top.reference"`
+			}{}
+
+			err = c.Unpack(&config, opts...)
+			assert.Error(t, err)
+		})
+	}
+}
+
+func TestDetectIndirectCyclicReference(t *testing.T) {
+	cfg := map[string]interface{}{
+		"a": map[string]interface{}{
+			"b": map[string]interface{}{
+				"c": "${a}",
+			},
+		},
+	}
+
+	config := struct {
+		A struct {
+			B struct {
+				C map[string]interface{}
+			}
+		}
+	}{}
+
+	c, err := NewFrom(cfg, opts...)
+	err = c.Unpack(&config, opts...)
+	if assert.Error(t, err) {
+		assert.Equal(t, "cyclic reference detected for key: 'a' accessing 'a.b'", err.Error())
+	}
+}
```

Comment:
as maps and structs do use different code-paths, tests should always be executed on a mix of structures.

```
tests := struct {
  name string
  config interface{}
}{
  { 
    "structs only", 
    &struct{
      ...
  },
  {
    "struct with pointers", ...
  }
  { "empty map": map[string]interface{}{} },

  ...
}

for _, test := range tests {
   ...
   err = c.Unpack(&test.config)
  ..
}

```

Reply:
I've did more of your comments in this PR, but I haven't push yet.

In the following code, when I try to unpack with test.config, I get `error: type 'struct { TopReference string "config:\"top.reference\"" }' is not supported on top level of config, only dictionary or list`
```golang
func TestDetectDirectCyclicReference(t *testing.T) {
	tests := []struct {
		title  string
		cfg    map[string]interface{}
		config interface{}
	}{
		{
			title: "direct reference on a struct",
			cfg: map[string]interface{}{
				"top.reference": "${top.reference}",
			},
			config: &struct {
				TopReference string `config:"top.reference"`
			}{},
		},
		{
			title: "direct compound reference on a struct",
			cfg: map[string]interface{}{
				"top.reference": "hello ${top.reference}",
			},
			config: &struct {
				TopReference string `config:"top.reference"`
			}{},
		},
		{
			title: "direct compound reference on an empty map",
			cfg: map[string]interface{}{
				"top.reference": "hello ${top.reference}",
			},
			config: map[string]interface{}{},
		},
		{
			title: "indirect compound reference on an empty map",
			cfg: map[string]interface{}{
				"a": map[string]interface{}{
					"b": map[string]interface{}{
						"c": "${a}",
					},
				},
			},
			config: map[string]interface{}{},
		},
	}

	for _, test := range tests {
		t.Run(test.title, func(t *testing.T) {
			c, err := NewFrom(test.cfg, opts...)

			err = c.Unpack(&test.config, opts...)
			assert.Error(t, err)
			fmt.Printf("error: %v\n", err)
		})
	}
}
```

Reply:
I think it is the `&test.config` being passed to `Unpack`. As the `test.config` already has a pointer to a struct when generating the array.

The error message might be misleading. I wonder if it shows the original type or the base-type (after iterating indirections like pointers and interfaces).

Reply:
I will do a bit more check, I will push my changes without the tests changes.
I have to pass &test.config, if I don't do that It will complaint that is not ptr.

I will if the error message is misleading.

Reply:
This might be due to the use of `map[string]interface{}{}`, instead of `&map[string]interface{}{}`. Have a look at the Unpack tests in `reify_test.go`. Something is weird about it complaining though.

Reply:
Same error after `&map[string]interface{}{}`, I did a quick look at the tests in `reify_test.go`, and the main difference is we loop through a slice of `[]interface{}` instead of looping through a slice of `[]struct` that has a field that we use as a target for `unpack`. 

We do some validation at https://github.com/ph/go-ucfg/blob/master/reify.go#L150 and we raise https://github.com/ph/go-ucfg/blob/ec8488a52542c0c51e42e8ea204dcaff400bc644/error.go#L189 we also raise the same error in merge. 

Running the same assertions outside of the `t.Run` and with a single variable as the unpack target works.

Reply:
I have update test with table based tests, https://github.com/elastic/go-ucfg/pull/97/files#diff-6902290fbccd8cbcdab6f13e5c7e26fdR72 I've rebased and cleanup.